File: 053_Simplify Path.cpp
/*
Simplify Path

Simplify an absolute Unix-style path.
*/

#include <bits/stdc++.h>
using namespace std;

string simplifyPath(string path) {
    vector<string> st;
    stringstream ss(path);
    string token;
    while (getline(ss, token, '/')) {
        if (token.empty() || token == ".") continue;
        if (token == "..") {
            if (!st.empty()) st.pop_back();
        } else st.push_back(token);
    }
    if (st.empty()) return "/";
    string res;
    for (auto &p: st) res += "/" + p;
    return res;
}

int main() {
    cout << "Simplify 1: " << simplifyPath("/home/") << endl; // /home
    cout << "Simplify 2: " << simplifyPath("/a/./b/../../c/") << endl; // /c
    cout << "Simplify 3: " << simplifyPath("/../") << endl; // /
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(n)

File: 053_Simplify Path.py
"""
Simplify Path

Simplify an absolute Unix-style path.
"""

def simplify_path(path: str) -> str:
    parts = []
    for token in path.split('/'):
        if not token or token == '.':
            continue
        if token == '..':
            if parts: parts.pop()
        else:
            parts.append(token)
    return '/' + '/'.join(parts)


if __name__ == '__main__':
    print('Simplify 1:', simplify_path('/home/'))
    print('Simplify 2:', simplify_path('/a/./b/../../c/'))
    print('Simplify 3:', simplify_path('/../'))

# Time Complexity: O(n)
# Space Complexity: O(n)
